### ISSUE TITLE
Macbook installation error

### ISSUE THREAD

**Issue by gdtm86** [ray-project/ray#461]
Hi, I'm following the instructions to install Ray on a macbook - http://ray.readthedocs.io/en/latest/install-on-macosx.html#installation-on-mac-os-x. 

So far I installed the dependencies successfully. But when I try to install Ray using 'python setup.py install --user', I'm getting build failures. 
```
/Users/gmedasani/software/anaconda_2_7/anaconda/include/python2.7/pyport.h:723:9: note: macro 'islower' defined here
#define islower(c) iswlower(btowc(c))
        ^
fatal error: too many errors emitted, stopping now [-ferror-limit=]
20 errors generated.
make[2]: *** [src/arrow/python/CMakeFiles/arrow_python_objlib.dir/common.cc.o] Error 1
make[1]: *** [src/arrow/python/CMakeFiles/arrow_python_objlib.dir/all] Error 2
make: *** [all] Error 2
Traceback (most recent call last):
  File "setup.py", line 46, in <module>
    license="Apache 2.0")
  File "/Users/gmedasani/software/anaconda_2_7/anaconda/lib/python2.7/distutils/core.py", line 151, in setup
    dist.run_commands()
  File "/Users/gmedasani/software/anaconda_2_7/anaconda/lib/python2.7/distutils/dist.py", line 953, in run_commands
    self.run_command(cmd)
  File "/Users/gmedasani/software/anaconda_2_7/anaconda/lib/python2.7/distutils/dist.py", line 972, in run_command
    cmd_obj.run()
  File "setup.py", line 13, in run
    subprocess.check_call(["../build.sh"])
  File "/Users/gmedasani/software/anaconda_2_7/anaconda/lib/python2.7/subprocess.py", line 541, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['../build.sh']' returned non-zero exit status 2
gurudharmatejas-MacBook-Pro:python gmedasani$
```

My Operating System version and Python version.

```
gurudharmatejas-MacBook-Pro:python gmedasani$ python --version
Python 2.7.12 :: Anaconda custom (x86_64)

gurudharmatejas-MacBook-Pro:python gmedasani$ system_profiler SPSoftwareDataType
Software:

    System Software Overview:

      System Version: macOS 10.12.4 (16E195)
      Kernel Version: Darwin 16.5.0
      Boot Volume: Macintosh HD
gurudharmatejas-MacBook-Pro:python gmedasani$

```

Libgcc version
```
gurudharmatejas-MacBook-Pro:python gmedasani$ conda install libgcc
Fetching package metadata .........
Solving package specifications: .

# All requested packages already installed.
# packages in environment at /Users/gmedasani/software/anaconda_2_7/anaconda:
#
libgcc                    4.8.5                         1
gurudharmatejas-MacBook-Pro:python gmedasani$
```

**Comment by robertnishihara** [comment]
Hi, it's possible we're running into problems compiling Apache Arrow. Can you try just compiling Arrow? You can use the following script.

```bash
mkdir build_arrow_temp
cd build_arrow_temp
git clone https://github.com/apache/arrow

mkdir arrow/cpp/build
cd arrow/cpp/build
git checkout b0863cb63d62ae7c4a429164e5a2e350d3c1f21a

cmake -DCMAKE_BUILD_TYPE=Release -DCMAKE_C_FLAGS="-g -O3 -march=native -mtune=native" -DCMAKE_CXX_FLAGS="-g -O3 -march=native -mtune=native" -DARROW_BUILD_TESTS=OFF -DARROW_PYTHON=on ..
make VERBOSE=1
```

Does that produce any errors? If you look through the logs, is it using the correct version of Python?

If that didn't work, does the above script work when you remove `-DARROW_PYTHON=on`? Note that if you rerun the script, you'll want to remove the files from the first run.

Can you attach the logs? Thanks!

@pcmoritz do you have other ideas?

**Comment by pcmoritz** [comment]
@gdtm86 Yeah, please try compiling error and also attach the full build output. I think there might have been more errors before the one you showed, which might give more insight. Also which library versions CMake found might help us to reproduce the problem.

**Comment by gdtm86** [comment]
Will give this a try once I get home as I was trying it on my personal laptop :).

**Comment by gdtm86** [comment]
Ok, gave this a try and the build failed with the -DARROW_PYTHON=on, but succeeded without the -DARROW_PYTHON=on. 

I attached both the logs. I verified the python version when I specified the -DARROW_PYTHON=on, and it is listed correctly. 

```
-- Found PythonInterp: /Users/gmedasani/software/anaconda_2_7/anaconda/bin/python (found version "2.7.12")
-- Found PythonLibs: /Users/gmedasani/software/anaconda_2_7/anaconda/lib/libpython2.7.dylib
-- Found NumPy: version "1.12.1" /Users/gmedasani/software/anaconda_2_7/anaconda/lib/python2.7/site-packages/numpy/core/include
```

[With-Python-ON-failure.txt](https://github.com/ray-project/ray/files/923224/With-Python-ON-failure.txt)
[Without-Python-ON-success.txt](https://github.com/ray-project/ray/files/923225/Without-Python-ON-success.txt)

**Comment by robertnishihara** [comment]
Ok, thanks! I'll file an issue with Arrow.

**Comment by robertnishihara** [comment]
See https://issues.apache.org/jira/browse/ARROW-826.

**Comment by pcmoritz** [comment]
@gdtm86 Can you try this:
```
mkdir build_arrow_temp
cd build_arrow_temp
git clone https://github.com/pcmoritz/arrow

mkdir arrow/cpp/build
cd arrow/cpp/build
git checkout cf59732a744eea8e3d736f59f65c5bc336331c05

cmake -DCMAKE_BUILD_TYPE=Release -DCMAKE_C_FLAGS="-g -O3 -march=native -mtune=native" -DCMAKE_CXX_FLAGS="-g -O3 -march=native -mtune=native" -DARROW_BUILD_TESTS=OFF -DARROW_PYTHON=on ..
make VERBOSE=1
```

and let me know if that fixes the problem for you?

**Comment by gdtm86** [comment]
Ok. that worked. Should I make any changes to the Ray build then?

**Comment by pcmoritz** [comment]
Great, let's get the changes merged into arrow and then you can go ahead and change the build to point to the fix in arrow.

**Comment by gdtm86** [comment]
Thank you Philipp.

**Comment by robertnishihara** [comment]
Great! Note that moving to the Arrow master will probably require addressing #444 and #447.

**Comment by pcmoritz** [comment]
That's a good point.

@gdtm86 If you are interested in helping out with these changes let us know that'd be great. Otherwise we can do it first and then get your patch merged!

**Comment by gdtm86** [comment]
@pcmoritz @robertnishihara are those changes something that need to be done in Arrow? or are there any related changes that need to be done in Ray?

**Comment by robertnishihara** [comment]
Those are changes that need to be done in Ray (in order to compile with a more recent version of Arrow). We are currently a few dozen commits behind the Arrow muster.

**Comment by gdtm86** [comment]
Sorry, I hit 'close' instead of comment. I can give this a try. I just opened the code in an IDE and looks like there are quite a lot of places where Arrow needs to be updated? Can you point where changes need to be made?

**Comment by pcmoritz** [comment]
Yes, all the changes should be in the files here:
https://github.com/ray-project/ray/tree/master/src/numbuf/cpp/src/numbuf
and here:
https://github.com/ray-project/ray/tree/master/src/numbuf/python/src/pynumbuf

The best way to do it I think is to change
https://github.com/ray-project/ray/blob/master/src/numbuf/thirdparty/download_thirdparty.sh
to replace https://github.com/apache/arrow/ with https://github.com/pcmoritz/arrow/
and b0863cb63d62ae7c4a429164e5a2e350d3c1f21a with cf59732a744eea8e3d736f59f65c5bc336331c05
and then fix the compile errors.

If you already built the project, you need to rf -rf ray/src/numbuf/thirdparty/arrow before calling setup.py, so the new version gets downloaded.

Let us know if you run into problems!

**Comment by gdtm86** [comment]
Ok, I was able to update the https://github.com/ray-project/ray/blob/master/src/numbuf/thirdparty/download_thirdparty.sh to include the changes you mentioned
```
if [ ! -d $TP_DIR/arrow ]; then
  git clone https://github.com/pcmoritz/arrow/ "$TP_DIR/arrow"
fi
cd $TP_DIR/arrow
git checkout cf59732a744eea8e3d736f59f65c5bc336331c05
```
Then compiled and ran into the arrow API issues in https://github.com/ray-project/ray/blob/master/src/numbuf/python/src/pynumbuf/adapters/python.cc. Here are the changes I made. 

Changed
```
  switch (arr->type()->type) {
```
to
```
  switch (arr->type()->id() ) {
```

And also changed the 'name' to 'name()' as shown below.
```
if (s->type()->child(0)->name() == "list") {
        return DeserializeList(l->values(), l->value_offset(index),
            l->value_offset(index + 1), base, tensors, result);
      } else if (s->type()->child(0)->name() == "tuple") {
        return DeserializeTuple(l->values(), l->value_offset(index),
            l->value_offset(index + 1), base, tensors, result);
      } else if (s->type()->child(0)->name() == "dict") {
        return DeserializeDict(l->values(), l->value_offset(index),
            l->value_offset(index + 1), base, tensors, result);
      } else {
        DCHECK(false) << "error";
      }
```

This helped me pass the arrow API errors. But I got a different error.

```
[ 93%] Built target plasma_manager
[ 94%] Building CXX object src/numbuf/CMakeFiles/numbuf.dir/python/src/pynumbuf/adapters/python.cc.o
make[2]: *** No rule to make target `../../../src/numbuf/thirdparty/arrow/cpp/build/release/libarrow_io.a', needed by `src/numbuf/libnumbuf.so'.  Stop.
make[2]: *** Waiting for unfinished jobs....
[ 95%] Building CXX object src/numbuf/CMakeFiles/numbuf.dir/__/common/lib/python/common_extension.cc.o
[ 96%] Linking CXX executable local_scheduler
[ 97%] Linking CXX executable manager_tests
[ 97%] Built target local_scheduler
[ 97%] Built target manager_tests
[ 98%] Linking CXX executable local_scheduler_tests
[ 98%] Built target local_scheduler_tests
make[1]: *** [src/numbuf/CMakeFiles/numbuf.dir/all] Error 2
make: *** [all] Error 2
Traceback (most recent call last):
  File "setup.py", line 46, in <module>
    license="Apache 2.0")
  File "/Users/gmedasani/software/anaconda_2_7/anaconda/lib/python2.7/distutils/core.py", line 151, in setup
    dist.run_commands()
  File "/Users/gmedasani/software/anaconda_2_7/anaconda/lib/python2.7/distutils/dist.py", line 953, in run_commands
    self.run_command(cmd)
  File "/Users/gmedasani/software/anaconda_2_7/anaconda/lib/python2.7/distutils/dist.py", line 972, in run_command
    cmd_obj.run()
  File "setup.py", line 13, in run
    subprocess.check_call(["../build.sh"])
  File "/Users/gmedasani/software/anaconda_2_7/anaconda/lib/python2.7/subprocess.py", line 541, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['../build.sh']' returned non-zero exit status 2
gurudharmatejas-MacBook-Pro:python gmedasani$
```

**Comment by pcmoritz** [comment]
Cool, the changes look good!

I think the error message is further up. Why don't you push the changes to your ray repo and post the branch here and I take a quick look?

**Comment by gdtm86** [comment]
Sure. Here is the link to my repo - https://github.com/gdtm86/ray/blob/master/src/numbuf/python/src/pynumbuf/adapters/python.cc

**Comment by robertnishihara** [comment]
Hmm.. is there an easy way for us to view the diff? If you fork the repository by clicking the GitHub fork button and make changes on top of that, the diff is easy to view,

**Comment by gdtm86** [comment]
Yeah. I messed that up. I forked the repo, but I made all the changes in the original repo I downloaded from here. Sorry about that. So I deleted my fork and created a new repo with updated code from my laptop. But I agree that fork will make it easier. I'll repeat the process and post the forked repo here. :)

**Comment by pcmoritz** [comment]
Btw, it is also possible to point an already cloned repo (from upstream) to your fork by running
```
git remote set-url origin https://github.com/gdtm86/ray
```
and then push it.

**Comment by gdtm86** [comment]
@pcmoritz - Thanks for that tip. Since the changes were minimal, I went ahead and did the old way :). Here is the repo - https://github.com/gdtm86/ray. It has the cmake-build-debug directory in it as well.

**Comment by pcmoritz** [comment]
I think I know what the problem is, the CMake script needs to be adapted a little. You can either make me a collaborator on your fork and I commit it directly, or I send you the patch, what do you prefer?

Also, let's get rid of the files generated by CMake please, because they will change a lot and it will make the history hard to follow if they are always checked in.

**Comment by gdtm86** [comment]
@pcmoritz Yes, I updated my .gitignore to ignore the ./idea and cmake-build-debug files. They are removed from my repo now. I just made you a collaborator.

**Comment by gdtm86** [comment]
@robertnishihara - you should be able to see the diff now. https://github.com/ray-project/ray/compare/master...gdtm86:master

**Comment by pcmoritz** [comment]
It builds now, but some boost library is missing:
```
ImportError: /home/pcmoritz/anaconda3/lib/python3.5/site-packages/ray-0.0.1-py3.5.egg/ray/core/src/numbuf/libnumbuf.so: undefined symbol: _ZN5boost9re_detail14verify_optionsEjNS_15regex_constants12_match_flagsE
```
I need to look some more into this...

**Comment by gdtm86** [comment]
Let me give this a try on my laptop.

**Comment by pcmoritz** [comment]
Oh and can you also create a PR to the Ray repo so the CI runs and we can track the changes?

**Comment by gdtm86** [comment]
I tried to build after pulling your changes in CMakesList.txt. But it still failed with some boost issues :(

```
[100%] Linking CXX shared library libnumbuf.so
Undefined symbols for architecture x86_64:
  "boost::basic_regex<char, boost::regex_traits<char, boost::cpp_regex_traits<char> > >::do_assign(char const*, char const*, unsigned int)", referenced from:
      __GLOBAL__sub_I_decimal.cc in libarrow.a(decimal.cc.o)
  "boost::re_detail_106300::get_mem_block()", referenced from:
      boost::re_detail_106300::perl_matcher<std::__1::__wrap_iter<char const*>, std::__1::allocator<boost::sub_match<std::__1::__wrap_iter<char const*> > >, boost::regex_traits<char, boost::cpp_regex_traits<char> > >::match_imp() in libarrow.a(decimal.cc.o)
      boost::re_detail_106300::perl_matcher<std::__1::__wrap_iter<char const*>, std::__1::allocator<boost::sub_match<std::__1::__wrap_iter<char const*> > >, boost::regex_traits<char, boost::cpp_regex_traits<char> > >::match_all_states() in libarrow.a(decimal.cc.o)
      boost::re_detail_106300::perl_matcher<std::__1::__wrap_iter<char const*>, std::__1::allocator<boost::sub_match<std::__1::__wrap_iter<char const*> > >, boost::regex_traits<char, boost::cpp_regex_traits<char> > >::match_startmark() in libarrow.a(decimal.cc.o)
      boost::re_detail_106300::perl_matcher<std::__1::__wrap_iter<char const*>, std::__1::allocator<boost::sub_match<std::__1::__wrap_iter<char const*> > >, boost::regex_traits<char, boost::cpp_regex_traits<char> > >::match_alt() in libarrow.a(decimal.cc.o)
      boost::re_detail_106300::perl_matcher<std::__1::__wrap_iter<char const*>, std::__1::allocator<boost::sub_match<std::__1::__wrap_iter<char const*> > >, boost::regex_traits<char, boost::cpp_regex_traits<char> > >::match_rep() in libarrow.a(decimal.cc.o)
      boost::re_detail_106300::perl_matcher<std::__1::__wrap_iter<char const*>, std::__1::allocator<boost::sub_match<std::__1::__wrap_iter<char const*> > >, boost::regex_traits<char, boost::cpp_regex_traits<char> > >::match_char_repeat() in libarrow.a(decimal.cc.o)
      boost::re_detail_106300::perl_matcher<std::__1::__wrap_iter<char const*>, std::__1::allocator<boost::sub_match<std::__1::__wrap_iter<char const*> > >, boost::regex_traits<char, boost::cpp_regex_traits<char> > >::match_set_repeat() in libarrow.a(decimal.cc.o)
      ...
  "boost::re_detail_106300::put_mem_block(void*)", referenced from:
      boost::re_detail_106300::perl_matcher<std::__1::__wrap_iter<char const*>, std::__1::allocator<boost::sub_match<std::__1::__wrap_iter<char const*> > >, boost::regex_traits<char, boost::cpp_regex_traits<char> > >::match_imp() in libarrow.a(decimal.cc.o)
      boost::re_detail_106300::perl_matcher<std::__1::__wrap_iter<char const*>, std::__1::allocator<boost::sub_match<std::__1::__wrap_iter<char const*> > >, boost::regex_traits<char, boost::cpp_regex_traits<char> > >::unwind_extra_block(bool) in libarrow.a(decimal.cc.o)
  "boost::re_detail_106300::verify_options(unsigned int, boost::regex_constants::_match_flags)", referenced from:
      boost::re_detail_106300::perl_matcher<std::__1::__wrap_iter<char const*>, std::__1::allocator<boost::sub_match<std::__1::__wrap_iter<char const*> > >, boost::regex_traits<char, boost::cpp_regex_traits<char> > >::match_imp() in libarrow.a(decimal.cc.o)
  "boost::re_detail_106300::raise_runtime_error(std::runtime_error const&)", referenced from:
      void boost::re_detail_106300::raise_error<boost::regex_traits_wrapper<boost::regex_traits<char, boost::cpp_regex_traits<char> > > >(boost::regex_traits_wrapper<boost::regex_traits<char, boost::cpp_regex_traits<char> > > const&, boost::regex_constants::error_type) in libarrow.a(decimal.cc.o)
  "boost::re_detail_106300::get_default_error_string(boost::regex_constants::error_type)", referenced from:
      boost::re_detail_106300::cpp_regex_traits_implementation<char>::error_string(boost::regex_constants::error_type) const in libarrow.a(decimal.cc.o)
  "boost::re_detail_106300::cpp_regex_traits_implementation<char>::transform_primary(char const*, char const*) const", referenced from:
      std::__1::__wrap_iter<char const*> boost::re_detail_106300::re_is_set_member<std::__1::__wrap_iter<char const*>, char, boost::regex_traits<char, boost::cpp_regex_traits<char> >, unsigned int>(std::__1::__wrap_iter<char const*>, std::__1::__wrap_iter<char const*>, boost::re_detail_106300::re_set_long<unsigned int> const*, boost::re_detail_106300::regex_data<char, boost::regex_traits<char, boost::cpp_regex_traits<char> > > const&, bool) in libarrow.a(decimal.cc.o)
  "boost::re_detail_106300::cpp_regex_traits_implementation<char>::transform(char const*, char const*) const", referenced from:
      std::__1::__wrap_iter<char const*> boost::re_detail_106300::re_is_set_member<std::__1::__wrap_iter<char const*>, char, boost::regex_traits<char, boost::cpp_regex_traits<char> >, unsigned int>(std::__1::__wrap_iter<char const*>, std::__1::__wrap_iter<char const*>, boost::re_detail_106300::re_set_long<unsigned int> const*, boost::re_detail_106300::regex_data<char, boost::regex_traits<char, boost::cpp_regex_traits<char> > > const&, bool) in libarrow.a(decimal.cc.o)
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [src/numbuf/libnumbuf.so] Error 1
make[1]: *** [src/numbuf/CMakeFiles/numbuf.dir/all] Error 2
make: *** [all] Error 2
Traceback (most recent call last):
  File "setup.py", line 46, in <module>
    license="Apache 2.0")
  File "/Users/gmedasani/software/anaconda_2_7/anaconda/lib/python2.7/distutils/core.py", line 151, in setup
    dist.run_commands()
  File "/Users/gmedasani/software/anaconda_2_7/anaconda/lib/python2.7/distutils/dist.py", line 953, in run_commands
    self.run_command(cmd)
  File "/Users/gmedasani/software/anaconda_2_7/anaconda/lib/python2.7/distutils/dist.py", line 972, in run_command
    cmd_obj.run()
  File "setup.py", line 13, in run
    subprocess.check_call(["../build.sh"])
  File "/Users/gmedasani/software/anaconda_2_7/anaconda/lib/python2.7/subprocess.py", line 541, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['../build.sh']' returned non-zero exit status 2
gurudharmatejas-MacBook-Pro:python gmedasani$
```

**Comment by pcmoritz** [comment]
Hm, these are related, my error is also due to boost_regex. Maybe @wesm has an idea what is going wrong here?

**Comment by xhochy** [comment]
This looks like you also need to add `libboost_regex.a` to the dependencies of `libnumbuf.so`. 

For future reference, it might be useful to run the build in this case with `make VERBOSE=1` so that the full, failed linker command is in the log.

**Comment by wesm** [comment]
I want to see if we can use `std::regex` instead of `boost::regex` as there was some OS X linker issue. I haven't had time to look into it

**Comment by pcmoritz** [comment]
@gdtm86 I implemented the solution that @xhochy suggested and pushed it to your repo. It works on my machine. Can you try it out on your Mac and create a pull request for ray so the CI gets run?

**Comment by gdtm86** [comment]
Thanks @pcmoritz. My build succeeded now and all the tests ran successfully.
```
79908:M 16 Apr 07:50:20.187 # User requested shutdown...
79908:M 16 Apr 07:50:20.187 # Redis is now ready to exit, bye bye...
ok

----------------------------------------------------------------------
Ran 32 tests in 78.447s

OK
gurudharmatejas-MacBook-Pro:ray gmedasani$
```

**Comment by gdtm86** [comment]
@pcmoritz. I'm creating a pull request now.

**Comment by gdtm86** [comment]
@pcmoritz I created the pull request and CI jobs are running now.

**Comment by robertnishihara** [comment]
Addressed by #464.
